plugins/nvme/fu-nvme-device.c

```diff
@@ -0,0 +1,330 @@
+}
+
+static gboolean
+fu_nvme_device_fw_download (FuNvmeDevice *self,
+			    guint32 addr,
+			    const guint8 *data,
+			    guint32 data_sz,
+			    GError **error)
+{
+	struct nvme_admin_cmd cmd = {
+		.opcode		= 0x11,
+		.addr		= 0x0, /* memory address of data */
+		.data_len	= data_sz,
+		.cdw10		= (data_sz >> 2) - 1,
+		.cdw11		= addr >> 2,
+	};
+	memcpy (&cmd.addr, &data, sizeof (gpointer));
+	return fu_nvme_device_submit_admin_passthru (self, &cmd, error);
+}
+
+static gboolean
+fu_nvme_device_parse_cns (FuNvmeDevice *self, const guint8 *buf, gsize sz, GError **error)
+{
+	guint8 fawr;
+	guint8 nfws;
+	guint8 s1ro;
+	g_autofree gchar *mn = NULL;
+	g_autofree gchar *sn = NULL;
+	g_autofree gchar *sr = NULL;
+
+	/* wrong size */
+	if (sz != FU_NVME_ID_CTRL_SIZE) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_FAILED,
+			     "failed to parse blob, expected 0x%04x bytes",
+			     (guint) FU_NVME_ID_CTRL_SIZE);
+		return FALSE;
+	}
+
+	/* get sanitiezed string from CNS -- see the following doc for offsets:
+	 * NVM-Express-1_3c-2018.05.24-Ratified.pdf */
+	sn = fu_nvme_device_get_string_safe (buf, 4, 23);
+	if (sn != NULL)
+		fu_device_set_serial (FU_DEVICE (self), sn);
+	mn = fu_nvme_device_get_string_safe (buf, 24, 63);
+	if (mn != NULL) {
+		fu_device_set_name (FU_DEVICE (self), mn);
+		fu_device_add_guid (FU_DEVICE (self), mn);
+	}
+	sr = fu_nvme_device_get_string_safe (buf, 64, 71);
+	if (sr != NULL)
+		fu_device_set_version (FU_DEVICE (self), sr);
+
+	/* firmware slot information */
+	fawr = (buf[260] & 0x10) >> 4;
+	nfws = (buf[260] & 0x0e) >> 1;
+	s1ro = buf[260] & 0x01;
+	g_debug ("nr fw slots: %u, slot1 r/o: %u", nfws, s1ro);
+	if (!fawr)
+		fu_device_add_flag (FU_DEVICE (self), FWUPD_DEVICE_FLAG_NEEDS_REBOOT);
+	return TRUE;
+}
+
+static guint
+fu_nvme_device_pci_slot_depth (GUdevDevice *device)
+{
+	g_autoptr(GUdevDevice) device_tmp = NULL;
+	device_tmp = g_udev_device_get_parent_with_subsystem (device, "pci", NULL);
+	if (device_tmp == NULL)
+		return 0;
+	for (guint i = 0; i < 0xff; i++) {
+		g_autoptr(GUdevDevice) parent = g_udev_device_get_parent (device_tmp);
+		if (parent == NULL)
+			return i;
+		g_set_object (&device_tmp, parent);
+	}
+	return 0;
+}
+
+gboolean
+fu_nvme_device_probe (FuNvmeDevice *self, GError **error)
+{
+	const gchar *vendor;
+	guint8 buf[FU_NVME_ID_CTRL_SIZE];
+	g_autoptr(GUdevDevice) parent = NULL;
+
+	/* look at the PCI depth to work out if in an external enclosure */
+	self->pci_depth = fu_nvme_device_pci_slot_depth (self->udev_device);
+	if (self->pci_depth <= 2)
+		fu_device_add_flag (FU_DEVICE (self), FWUPD_DEVICE_FLAG_INTERNAL);
+
+	/* get and parse CNS */
+	if (!fu_nvme_device_identify_ctrl (self, buf, error)) {
+		g_prefix_error (error, "failed to identify %s: ",
+				fu_device_get_platform_id (FU_DEVICE (self)));
+		return FALSE;
+	}
+	if (!fu_nvme_device_parse_cns (self, buf, sizeof(buf), error))
+		return FALSE;
+
```

Comment:
I think it would be good to add a function that at least dumps the vendor block into debug output until we know what to do with it.

Reply:
All 1024 bytes of it?

Reply:
Maybe keep the printing of it behind `FU_NVME_VERBOSE` or similar.  I think it would be good when we have some parsing methods to try to throw at it though to have the plumbing around it.

Reply:
Done.